ogr/ogrsf_frmts/miramon/ogrmiramondatasource.cpp

```diff
@@ -0,0 +1,310 @@
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE. 
+ ****************************************************************************/
+
+#include "ogrmiramon.h"
+
+/****************************************************************************/
+/*                          OGRMiraMonDataSource()                          */
+/****************************************************************************/
+OGRMiraMonDataSource::OGRMiraMonDataSource()
+    : papoLayers(nullptr), nLayers(0), pszRootName(nullptr), pszDSName(nullptr),
+      bUpdate(false)
+
+{
+    MMMap.nNumberOfLayers = 0;
+    MMMap.fMMMap = nullptr;
+}
+
+/****************************************************************************/
+/*                         ~OGRMiraMonDataSource()                          */
+/****************************************************************************/
+
+OGRMiraMonDataSource::~OGRMiraMonDataSource()
+
+{
+    for (int i = 0; i < nLayers; i++)
+        delete papoLayers[i];
+    CPLFree(papoLayers);
+    CPLFree(pszDSName);
+    CPLFree(pszRootName);
+
+    if (MMMap.fMMMap)
+        VSIFCloseL(MMMap.fMMMap);
+}
+
+/****************************************************************************/
+/*                                Open()                                    */
+/****************************************************************************/
+
+int OGRMiraMonDataSource::Open(const char *pszFilename, VSILFILE *fp,
+                               const OGRSpatialReference *poSRS, int bUpdateIn,
+                               CSLConstList papszOpenOptionsUsr)
+
+{
+    bUpdate = CPL_TO_BOOL(bUpdateIn);
+
+    OGRMiraMonLayer *poLayer = new OGRMiraMonLayer(
+        pszFilename, fp, poSRS, bUpdate, papszOpenOptionsUsr, &MMMap);
+    if (!poLayer->bValidFile)
+    {
+        delete poLayer;
+        return FALSE;
+    }
+    papoLayers = static_cast<OGRMiraMonLayer **>(CPLRealloc(
+        papoLayers,
+        (size_t)(sizeof(OGRMiraMonLayer *) * ((size_t)nLayers + (size_t)1))));
+    papoLayers[nLayers] = poLayer;
+    nLayers++;
+
+    if (pszDSName)
+    {
+        const char *pszExtension = CPLGetExtension(pszDSName);
+        if (!EQUAL(pszExtension, "pol") && !EQUAL(pszExtension, "arc") &&
+            !EQUAL(pszExtension, "pnt"))
+        {
+            strcpy(
```

Comment:
how can we reach that code as the Identify() function has already checked extensions ? Isn't it dead code ?
And it is highly unusual for a Open() code to have to write a file.

Reply:
There is a possibility of empty extension (perhaps xxx.yyy.aaa but aaa is not any of the three I mention) and -f MiraMonVector at command line to indicate that the result is a folder a file with name xxx.yyy.aaa.mmm is created to be opened easy with MiraMon software. 
![2024-04-03 221320](https://github.com/AbelPau/gdal/assets/92721356/6a0031d3-7379-496a-9afd-a538374539cf)

Reply:
> There is a possibility of empty extension

well, now that OGRMiraMonDriverOpen() relies on OGRMiraMonDriverIdentify(), the possibility of an empty extension isn't taken into account. Let's keep that like that. If there is a need to change this, this can be done after the driver has been merged, as follow-up pull requests. We should really try to keep changes in this branch to the bare minimum otherwise this is going to be unmanageable